File: tests/matching.test.ts
import { describe, expect, it } from "vitest";

import { __testing } from "@/lib/pokemontcg";
import type { Card, ScanParse } from "@/lib/types";

const { scoreCard, normaliseNumber, similarity } = __testing;

function card(overrides: Partial<Card> = {}): Card {
  return {
    id: "sv3pt5-54",
    name: "Charizard ex",
    number: "054",
    printedTotal: 165,
    setId: "sv3pt5",
    setName: "151",
    setSeries: "Scarlet & Violet",
    setReleaseDate: "2023/09/22",
    rarity: "Double Rare",
    images: { small: "", large: "" },
    tcgplayerUrl: null,
    cardmarketUrl: null,
    rawTcgplayer: null,
    rawCardmarket: null,
    ...overrides,
  };
}

function scan(overrides: Partial<ScanParse> = {}): ScanParse {
  return {
    name: null,
    number: null,
    printedTotal: null,
    setCode: null,
    rawText: "",
    confidence: 0,
    ...overrides,
  };
}

describe("normaliseNumber", () => {
  it("ignores leading zeros", () => {
    expect(normaliseNumber("054")).toBe(normaliseNumber("54"));
  });

  it("keeps a subset prefix", () => {
    expect(normaliseNumber("TG012")).toBe("TG12");
  });
});

describe("similarity", () => {
  it("scores identical strings as 1", () => {
    expect(similarity("charizard", "charizard")).toBe(1);
  });

  it("stays high through a single OCR error", () => {
    expect(similarity("charizard", "chariza rd".replace(" ", ""))).toBeGreaterThan(0.9);
  });

  it("scores unrelated names low", () => {
    expect(similarity("charizard", "blastoise")).toBeLessThan(0.4);
  });
});

describe("scoreCard", () => {
  it("scores a full match near 1", () => {
    const { score } = scoreCard(
      card(),
      scan({ name: "Charizard ex", number: "054", printedTotal: 165 }),
    );
    expect(score).toBeGreaterThan(0.95);
  });

  it("still clears the candidate threshold on the number alone", () => {
    const { score } = scoreCard(card(), scan({ number: "054", printedTotal: 165 }));
    expect(score).toBeGreaterThan(0.35);
  });

  it("scores a wrong card below the threshold", () => {
    const { score } = scoreCard(
      card({ name: "Blastoise ex", number: "009" }),
      scan({ name: "Charizard ex", number: "054" }),
    );
    expect(score).toBeLessThan(0.35);
  });

  it("explains why it matched", () => {
    const { reasons } = scoreCard(
      card(),
      scan({ name: "Charizard ex", number: "054", printedTotal: 165 }),
    );
    expect(reasons).toContain("Exact name match");
    expect(reasons).toContain("Card number 054");
  });
});

File: android/core/src/test/kotlin/com/paladex/ex/core/MatchingTest.kt
package com.paladex.ex.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class MatchingTest {

    private fun card(name: String = "Charizard ex", number: String = "054") = Card(
        id = "sv3pt5-54",
        name = name,
        number = number,
        printedTotal = 165,
        setId = "sv3pt5",
        setName = "151",
        setSeries = "Scarlet & Violet",
        rarity = "Double Rare",
    )

    @Test
    fun `number comparison ignores leading zeros`() {
        assertEquals(
            PokemonTcgClient.normaliseNumber("54"),
            PokemonTcgClient.normaliseNumber("054"),
        )
    }

    @Test
    fun `number comparison keeps a subset prefix`() {
        assertEquals("TG12", PokemonTcgClient.normaliseNumber("TG012"))
    }

    @Test
    fun `identical strings score 1`() {
        assertEquals(1.0, PokemonTcgClient.similarity("charizard", "charizard"))
    }

    @Test
    fun `similarity stays high through a single OCR error`() {
        assertTrue(PokemonTcgClient.similarity("charizard", "charizerd") > 0.85)
    }

    @Test
    fun `unrelated names score low`() {
        assertTrue(PokemonTcgClient.similarity("charizard", "blastoise") < 0.4)
    }

    @Test
    fun `scores a full match near 1`() {
        val scan = ScanParse(name = "Charizard ex", number = "054", printedTotal = 165)
        assertTrue(PokemonTcgClient.scoreCard(card(), scan).score > 0.95)
    }

    @Test
    fun `still clears the candidate threshold on the number alone`() {
        val scan = ScanParse(number = "054", printedTotal = 165)
        assertTrue(PokemonTcgClient.scoreCard(card(), scan).score > 0.35)
    }

    @Test
    fun `scores a wrong card below the threshold`() {
        val scan = ScanParse(name = "Charizard ex", number = "054")
        val other = card(name = "Blastoise ex", number = "009")
        assertTrue(PokemonTcgClient.scoreCard(other, scan).score < 0.35)
    }

    @Test
    fun `explains why it matched`() {
        val scan = ScanParse(name = "Charizard ex", number = "054", printedTotal = 165)
        val reasons = PokemonTcgClient.scoreCard(card(), scan).reasons

        assertTrue("Exact name match" in reasons)
        assertTrue("Card number 054" in reasons)
    }
}
